File: frontend/src/types/widget.ts
/**
 * GSAD Dashboard Widget Type Definitions
 * Grid Specification v1.1
 */

/** Priority levels 1-5 (1 = core, 5 = optional) */
export type PriorityLevel = 1 | 2 | 3 | 4 | 5;

/** Widget category for grouping and filtering */
export type WidgetCategory =
  | 'geospatial'
  | 'ai'
  | 'strategic'
  | 'news'
  | 'webcams'
  | 'intelligence'
  | 'markets'
  | 'technology'
  | 'risk'
  | 'crypto';

/** Cross-linking index — every widget event must carry these fields */
export interface CrossLinkPayload {
  eventId: string;
  countryCode: string;
  severityScore: number;
  timestamp: string; // ISO-8601
  categoryTags: string[];
}

/** Persistent grid position for layout storage */
export interface GridPosition {
  colStart: number;
  colSpan: number;
  rowStart: number;
  rowSpan: number;
}

/** Data source descriptor attached to each widget */
export interface DataSourceConfig {
  provider: string;
  endpoint?: string;
  apiKeyEnvVar?: string;
  refreshIntervalMs: number;
  fallback?: string;
}

/** Full widget definition */
export interface WidgetDefinition {
  id: string;             // e.g. "GSAD-01"
  name: string;           // human-readable
  category: WidgetCategory;
  priority: PriorityLevel;
  defaultPosition: GridPosition;
  dataSources: DataSourceConfig[];
  features: string[];
  description: string;
}

/** Runtime widget state */
export interface WidgetState {
  id: string;
  isLoading: boolean;
  hasError: boolean;
  errorMessage?: string;
  lastUpdated?: string;
  isExpanded: boolean;
  isFocused: boolean;
}

/** Layout persistence shape (matches spec §9.2) */
export interface PersistedLayout {
  widget_id: string;
  col_start: number;
  col_span: number;
  row_start: number;
  row_span: number;
}

/** Saved layout preset */
export interface LayoutPreset {
  id: string;
  name: string;
  description: string;
  widgets: PersistedLayout[];
  createdAt: string;
  updatedAt: string;
}

File: backend/src/models/widget.rs
use serde::{Deserialize, Serialize};

/// Mirrors the TypeScript `WidgetDefinition` type.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WidgetDefinition {
    pub id: String,
    pub name: String,
    pub category: String,
    pub priority: u8,
    pub default_position: GridPosition,
    pub data_sources: Vec<DataSourceConfig>,
    pub features: Vec<String>,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GridPosition {
    pub col_start: u32,
    pub col_span: u32,
    pub row_start: u32,
    pub row_span: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DataSourceConfig {
    pub provider: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_key_env_var: Option<String>,
    pub refresh_interval_ms: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fallback: Option<String>,
}

/// Cross-linking payload carried by every event.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrossLinkPayload {
    pub event_id: String,
    pub country_code: String,
    pub severity_score: f64,
    pub timestamp: String,
    pub category_tags: Vec<String>,
}

/// Layout persistence (matches spec §9.2).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersistedLayout {
    pub widget_id: String,
    pub col_start: u32,
    pub col_span: u32,
    pub row_start: u32,
    pub row_span: u32,
}
